Conversion of all blocks to Block API

Bottersnike opened the pull request:
This is a WIP PR, and not ready for merge. As part of the MVC goal, separation of blocks out of the monolithic logo.js file is a high priority. Conversion to the proposed API is as follows:

1. Determine if a new file is needed for the block. If not, skip to (6)
2. Make a new file in `js/blocks/`, named something sensible
3. Add that file to `MUSICBLOCKS_EXTRAS` in `js/activity.js`
4. Create a setup function in your new file at the end, named something sensible
5. Call that function in `js/basicblocks.js`
6. Create a new class. All demonstrations extend from `BaseBlock`, however blocks can extend each other (for example, the default case block could have been made to extend the case block).
7. Locate the existing declaration in `js/basicblocks.js`. The argument passed to `new ProtoBlock` should be passed to the super call, the palette assignment replaced with `this.setPalette`, the `protoBlockDict` assignment discarded, and the other lines can be used verbatim, albeit replacing `newblock` with `this. If there is a `beginerMode && !beginne.....` check, it can be discarded, as `BaseBlock` automatically performs that check on all blocks.
8. Add a call to `new MyNewBlock().setup();` in the setup function
9. For arg blocks, define a function `arg` on the block. The currently passed arguments are `(logo, turtle, blk, receivedArg)`.
10. For flow blocks, define a function `flow` on the block. The currently passed arguments are `(args, logo, turtle, blk, receivedArg)` however trailing arguments can be omitted if not needed (thanks, JavaScript!).
11. Locate the logic for the block in `logo.js`. Move it into either of the two functions from (9) and (10). Remember to replace `that.` with `logo.`. For arg blocks, calls to `that.blocks.blockList[blk].value = ...` should be replaced with `return ...`.
12. In the case of flow blocks, the return value should be `[childFlow, childFlowCount]`, or `[]` if they are unchanged, so changes to these variables should be checked for and changed to return instead.
12. If the block is a flow block, the comment containing `----- ADD SMART BLOCK CLASSES HERE -----` should be located and the block ID added to the case stack. Argument blocks are automatically detected.
13. If the block is a macro block, all code relating to it should be removed from `macros.js` and the macro definition moved to `this.makeMacro((x, y) => [....])`. See `SwitchBlock` for an example of this.

Ideally, once all blocks are ported over, the case stack in (12) will contain every single block, and there will be no other cases, at which point the switch can be entirely removed.

Bottersnike commented:
The above commit adds a single method to `BaseBlock`, named `formBlock`. This function takes a JSON-like object describing the visual appearance of the block, the arguments it takes, etc.. The format of that object is as follows:

* `name`: The display name on the block. Typically `_('...')`. If this is omitted or an empty string, a second argument of `false` should be passed to `formBlock` to ensure it doesn't attempt to resize the block to fit the (non-existent) text.
* `flows`
  * `top`: How the top of the block should connect. One of `true`, `false`, or `'cap'`. `'cap'` is used for blocks like start, to produce the spike.
  * `bottom`: How the bottom of the block should connect. One of `true`, `false`, or `'tail'`. An example of tail can be seen in the above commit used in the break (stop) block.
  * `left`: How the left of the block should connect. This is just a boolean.
  * `type`: What sort of flow does this block use. This is one of `'flow'`, `'arg'`, `'value'`, or `null`. Flow blocks are the most common, with arg being used for blocks like do and calculate. `value` is used for blocks such as number, and a null value indicates that the block is just a plain ol' block with nothing special.
  * `labels`: The labels to be displayed for each flow branch. This list is also used to determine how many flows a block has. Empty strings can be used when an unlabelled flow is desired.
* `args`: The number of arguments this block takes (right-hand side). Special cases for this value are `'onebool'` and `'twobool'`, which are fairly self-explanatory.
* `argTypes`: The type for each argument. Default is `'numberin'`, but an example of this can be see in in the case block, where that is change to `'anyin'`.
* `argLabels`: A list of labels to be applied to the arguments.
* `argDefaults`: A list of default values for arguments.

The call to `formBlock` will attempt to further call `adjustWidthToLabel`, however this behaviour can be suppressed by passing a falsey value as the second argument. There is currently no way to define the left-hand output as a boolean, however this should just be a case of adding a third option of `'bool'` to `flows.left`, however that is something to work out when actually converting a block that requires that option.

@walterbender

walterbender commented:
there is one corner case where the default block name can be overridden.

Bottersnike commented:
Making a note so nothing gets merged while broken, there's currently a race condition present. If any of the blocks attempt to load before `BlockBase.js`, the following error occurs and blocks cease to load:

![image](https://user-images.githubusercontent.com/7524553/71547954-c252cf80-299e-11ea-86d2-4323d6b0f61a.png)

The fix for this is just wrapping all block class definition within requirejs handlers, and likewise put the definition for `BaseBlock` in a provider handler. For now the fix is just reload the page, as it is far more common for this race condition to load `BaseBlock` first.

Bottersnike commented:
Having just ported the action blocks to the API, it'd be interesting to have a huge MB project that made use of every single block, for testing. Is there such a project hanging around?

Side note: It's looking like the rate of porting is going to be about one palette per day, possible two per day at a stretch.

walterbender commented:
I will  make you a uber project for testing.

walterbender commented:
[uber.zip](https://github.com/sugarlabs/musicblocks/files/4008509/uber.zip)

Not quite all the blocks yet, but most of them. Fixed quite a few little bugs in the process of making this.

Bottersnike commented:
For the graphics palette, as the others, I performed a straight port. In the clear block, at https://github.com/sugarlabs/musicblocks/pull/1994/files#diff-2b77c9084b6f412a8075de688c2a396aR113, the line to actually perform the clear is commented out, which doesn't match up with the translation description a few lines above. I've kept this as-is, but is this intentional?

walterbender commented:
@Bottersnike I am confused re clear. Where is the source of the commented code? I don't see it here:

https://github.com/sugarlabs/musicblocks/blob/master/js/logo.js#L2440

Bottersnike commented:
Aah you absolutely right. I accidentally yoinked the definition from down at https://github.com/sugarlabs/musicblocks/blob/master/js/logo.js#L9291 🤦.

walterbender commented:
Oh. I see. 
https://github.com/sugarlabs/musicblocks/blob/master/js/logo.js#L9291

That is in the playback loop. That is code specific to running inside of a note. There is a bug (I cannot find the issue for some reason) when using clear inside a note.

walterbender commented:
Is this at a point where I can help with testing?

Bottersnike commented:
Currently the only palettes left to convert are Widgets, Drum, Volume, Ornament, Tone, Intervals, Pitch, and I'm about half-way though Meter right now. In theory, all the other palettes should be 100% converted and testable, and the non-converted ones should play along nicely.

walterbender commented:
OK. I'll dig in to it.

walterbender commented:
Mostly seems to be working. I tried running a pretty ambitious project (Crab Canon Plot) and part way through it fell very far behind. Then it caught up again. Seems to be a consistent error. Not sure yet of the root cause.

walterbender commented:
I isolated the problem to the silence block (when interacting with the graphics).

Bottersnike commented:
Hmm. The silence block seems like a peculiar cause. The flow function is just

```js
if (logo.inNoteBlock[turtle].length > 0) {
    logo.notePitches[turtle][last(logo.inNoteBlock[turtle])].push('rest');
    logo.noteOctaves[turtle][last(logo.inNoteBlock[turtle])].push(4);
    logo.noteCents[turtle][last(logo.inNoteBlock[turtle])].push(0);
    logo.noteHertz[turtle][last(logo.inNoteBlock[turtle])].push(0);
    logo.noteBeatValues[turtle][last(logo.inNoteBlock[turtle])].push(logo.beatFactor[turtle]);
    logo.pushedNote[turtle] = true;
}
```

which seems like an odd thing to be causing it to fall behind. It's also one of the blocks that has little-to-no modification from `logo.js` (just replacing that->logo).

walterbender commented:
I tried to dig a bit deeper into the problem I was having with the rest block (interacting with deltapitch and mynotevalue). I reproduced it repeatedly the other day but can no longer reproduce it. Not sure what changed, but presumably it was due to some funny state I had gotten my browser into???

walterbender commented:
@Bottersnike seems you are almost done. I can cherry pick (more migrate) the things that changed since you started -- a couple of new blocks and a couple of other small changes in logo.js before we do the merge... but it is all working well in my tests.

Bottersnike commented:
The main thing I foresee being an issue currently is the occasional https://github.com/sugarlabs/musicblocks/pull/1994#issuecomment-569440269. This is due to the fact requirejs loads all files at once, and I can only really think of two solutions:

1. Wrap each ...Blocks.js file in a `require` call, and explicitly define the setup function onto `window.`
2. Move the class definitions for each block into inside the setup function, where the existence of the base classes is guaranteed at call-time.

Of the two, I'm strongly leaning towards 2 because there's little to no extra boilerplate (just an extra indentation level), but both seem a little messy. Any ideas for other solutions?

walterbender commented:
Nothing else comes to mind, but I will investigate

Bottersnike commented:
@walterbender with those commits, that should be every single block and block macro converted. I think most of `turtledefs.js` can also be merged in, but I have a feeling that'd be better suited to its own entire PR. Testing wise, one important thing I've been unfortunately lax in testing is the differences between the eng and jp versions of the site, but that should just be a case of having the live site and this PR side-by-side and comparing them.

walterbender commented:
I should have time over the weekend to pull in the other relevant patches and do the side-by-side. And hopefully do a merge. Thanks for doing this. Should make it a much nicer experience for everyone going forward.

walterbender commented:
Not sure what happened with the latest commits, but I can no longer get this to work: in the main loop in logo.js it does not recognize any blocks. L2034 for every block.

walterbender commented:
The problem is with this commit:
commit 94caa8a25ba57e1f78f3e6b1efc8f90913b4b3ef

The changes you made in logo.js to get ride of all of the case statements doesn't work.

Bottersnike commented:
Aargh you're absolutely right! I forgot to replace the `if` after it with `else if` 🤦‍♂

walterbender commented:
@Bottersnike I think I have all of the relevant changes merged (e.g., distance block, etc.) and I fixed a few odds and ends. 

Something not quite right about boxes -- when you save a project with a box1 block, it is not restored properly. Investigating. 

Otherwise, I think we are good to go. Really looking forward to merging this.

walterbender commented:
Fixed the problem with boxes. But I don't know how it could be working in master. Very strange.

walterbender commented:
I pushed my edits to https://github.com/sugarlabs/musicblocks/tree/Bottersnike-block-class-edits

@Bottersnike @pikurasa please test for regressions before I merge to master

walterbender commented:
Seems to be a regression with arg blocks. Investigating.

walterbender commented:
Fixed a bunch more little bugs. Please do test if you get a chance. I think we are very close.

walterbender commented:
Some issues with the Loudness block (and the label on the Mouse button block).

walterbender commented:
Fixed the sensor problems.
Next problem: some blocks don't auto-resize any more (and sometimes the vspace blocks are not inserted)

pikurasa commented:
I do not know exactly what I should be focused on as I test, but the synth sounds do not seem to be initiating correctly for my tests. (They all play the default synth)

Bottersnike commented:
Huge apologies for the radio silence. Do you know if there are specific blocks are aren't auto-resizing? When initially testing I found that if-then-else blocks were having strange expansion issues, but they appeared to already exist prior to making any edits to the code. The insertion (or lack thereof) of vpsace blocks is a bit of an odd one. If it's on macro expansion there's a high chance I managed to delete them without realising when expanding the lists out into multiple lines. If they're being inserted and removed automatically during auto-resizing, my guess is there's an issue in the calculation of `this.size` in `BaseBlock`. Unfortunately, I'm not especially sure on the rules related to vpsace insertion.

In terms of the merges, most things seem to work. `RhythmBlockPaletteBlocks.js` had some strange scoping, which I've fixed. `PitchBlocks.js`, line 1580 appears to reference `transposition` which was originally being defined in `case 'pitch':` and `case 'steppitch':` and I think some funky scoping was letting it be accessed elsewhere. Not sure how to fix the latter.

@pikurasa I think what's happening there is that the request to load the samples is timing out. I'm able to reproduce the same results when I clear my browser cache and reload (forcing it to download all the samples again). I'm not sure how best to fix this, as it appears to be due to the shear amount of data that MB is having to pull from the server on page load. It may be possible to change it to only load a sample's data when a block is actually placed that references that sample, but I haven't looked into it and I'm not sure how practical it would be.

The changes on sugar/mb/Bottersnike-block-class-edits are now hopefully merged back into this PR

walterbender commented:
Line 103 of BaseBlock:        this.generator = function () {

Doesn't it needs o take an arg for how much to expand the clamp?

walterbender commented:
dot block is not working...

walterbender commented:
Ah. arguments. Modern JavaScript. What a concept!!!

I figured out the problem with basic two-arg blocks (we need to call setExpand rather than setClamp in those cases. Next I need to fix the two-arg math operators.

walterbender commented:
I have expanding working now :)

walterbender commented:
@Bottersnike I have a few more tweaks -- dot block and some missing refresh somewhere. The other blocker before we merge is the requirejs issue. It seems to hit me about 1 in 10 sessions... a bit too often to ignore.

walterbender commented:
Dot block works now :)

walterbender commented:
On second thought, I think I will merge now so as to up the number of people who will be able to help with uncovering any additional bugs. We can open a separate ticket re requirejs

walterbender commented:
@Bottersnike this is a great contribution. Thank you.

aviral243 commented:
After this PR when I fetched upstream and merged it, I'm getting errors as shown in the screenshot.

![Screenshot from 2020-01-31 13-48-39](https://user-images.githubusercontent.com/44023445/73523321-aa082180-4430-11ea-8b0b-524c90ac6729.png)
 Is this due to this PR or is something wrong with my local copy? 
@walterbender  Please help

Bottersnike commented:
@aviral243 those errors are normal. MusicBlocks relies on a large number of global variables. These cause eslint to get very annoyed :). One of the things that I think I accidentally committed was my `.eslintrc` file, which you're VSC has detected. Rather than the errors being new, it's just that your editor previously wasn't checking at all. Delete the `.eslintrc` file and it should go back to ignoring all of them.

aviral243 commented:
Yeah. Back to normal. Thanks a lot @Bottersnike .

walterbender commented:
FYI,help, the parameter updater, and block setter are now part of the Block API.

bartekpacia commented:
@Bottersnike I'm in awe of this wonder.
